### PR TITLE
NEON metadata ingest pipeline

### PR DESCRIPTION
**Pull request by sujaypatil96**
This PR is a work in progress that seeks to address https://github.com/microbiomedata/nmdc-runtime/issues/249

### CODE REVIEW

File: nmdc_runtime/site/translation/neon_translator.py [hunk 1]
```diff
@@ -0,0 +1,757 @@
...
+                mms_metagenomeDnaExtraction.laboratoryName,
+                mms_metagenomeDnaExtraction.collectDate,
+                mms_metagenomeDnaExtraction.processedDate,
+                mms_metagenomeSequencing.sequencingFacilityID,
+                mms_metagenomeSequencing.ncbiProjectID
+            FROM mms_metagenomeSequencing 
+            LEFT JOIN mms_metagenomeDnaExtraction ON mms_metagenomeDnaExtraction.dnaSampleID = mms_metagenomeSequencing.dnaSampleID
+        """
+        library_preparation_table = pd.read_sql_query(query, self.conn)
+        omics_processing_table = pd.read_sql_query(query, self.conn)
```
Comment (pkalita-lbl) [inline nmdc_runtime/site/translation/neon_translator.py:576]: Duplicate query?

File: nmdc_runtime/site/translation/neon_translator.py [hunk 2]
```diff
@@ -0,0 +1,757 @@
...
+    ) -> nmdc.OmicsProcessing:
+        processing_institution = None
+        sequencing_facility = self._get_value_or_none(
+            omics_processing_row, "sequencingFacilityID"
+        )
+        if sequencing_facility is not None:
+            if re.search("Battelle", sequencing_facility, re.IGNORECASE):
+                processing_institution = "Battelle"
+            elif re.search("Argonne", sequencing_facility, re.IGNORECASE):
+                processing_institution = "ANL"
```
Comment (pkalita-lbl) [inline nmdc_runtime/site/translation/neon_translator.py:342]: This appears a few times. Consider extracting to separate method.

File: nmdc_runtime/site/graphs.py [hunk 3]
```diff
@@ -146,3 +152,19 @@ def ingest_metadata_submission():
     database = translate_portal_submission_to_nmdc_schema_database(metadata_submission)
     run_id = submit_metadata_to_db(database)
     poll_for_run_completion(run_id)
+
+@graph
+def ingest_neon_metadata():
```
Comment (pkalita-lbl) [inline nmdc_runtime/site/graphs.py:157]: I made a change to the submission portal ingestion pipeline that must have gone in after this branch was created because I don't see it here. But if you look in `main` I made two different graphs `translate_metadata_submission_to_nmdc_schema_database` -- which is like the "dry run" version, it just does the translation, validates it, and stores the output -- and `ingest_metadata_submission` -- which actually submits the translation result to Mongo. You might want to consider the same pattern here.

**Review by pkalita-lbl — COMMENTED** [review]
It's a little hard to follow exactly what all the SQL queries are doing, but I trust based on your validation and testing that they're doing the right thing. Overall though this looks like it's going in the right direction.

File: nmdc_runtime/site/ops.py [hunk 4]
```diff
@@ -711,3 +714,87 @@ def export_json_to_drs(
...
+                for file in data_files["data"]["files"]:
+                    if table_name in file["name"] and "expanded" in file["name"]:
+                        current_df = pd.read_csv(file['url'])
+                        df = pd.concat([df, current_df], ignore_index=True)
+        df_dict[table_name] = df
+
+    return df_dict
+
+@op(required_resource_keys={"neon_api_client"})
+def neon_sls_data_by_product(
```
[the remark this answers is not in the document]
Comment (sujaypatil96) [reply nmdc_runtime/site/ops.py:755]: The difference is in the dict/list of values that each of those `@op` methods is capturing through the `@repository` _interface_.

The config schema for `neon_mms_data_by_product` looks like:

```python
"mms_data_product": {
  "product_id": "DP1.10107.001",
  "product_tables": "mms_metagenomeDnaExtraction, mms_metagenomeSequencing",
}
```

The config schema for `neon_mms_data_by_product` looks like:

```python
"sls_data_product": {
  "product_id": "DP1.10086.001",
  "product_tables": "sls_metagenomicsPooling, sls_soilCoreCollection, sls_soilChemistry, sls_soilMoisture, sls_soilpH, ntr_externalLab, ntr_internalLab",
}
```

File: nmdc_runtime/site/translation/neon_translator.py [hunk 5]
```diff
@@ -0,0 +1,757 @@
...
+from nmdc_schema import nmdc
+from nmdc_runtime.site.translation.translator import Translator
+from nmdc_runtime.site.util import get_basename
+
+
+class NeonDataTranslator(Translator):
+    def __init__(self, mms_data: dict, sls_data: dict, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+
+        self.conn = sqlite3.connect("neon.db")
```
[the remark this answers is not in the document]
Comment (sujaypatil96) [reply nmdc_runtime/site/translation/neon_translator.py:16]: Oh yes, I'll read up more on different db caching/storage options like `:memory:`.

File: nmdc_runtime/site/translation/neon_translator.py (hunk 1 above)
Comment (sujaypatil96) [reply nmdc_runtime/site/translation/neon_translator.py:597]: It's the same starting query from which we'll be creating the `LibraryPreparation` records and the `OmicsProcessing` records.

File: nmdc_runtime/site/graphs.py (hunk 3 above)
Comment (sujaypatil96) [reply nmdc_runtime/site/graphs.py:157]: Yes, for sure, that's a good pattern to follow, can choose to run either pipeline based on the action that the user wants to accomplishm I'll split out the "dry run" and actual submission to Mongo graphs into separate methods.

I'm curious though, this pattern hasn't been applied to the gold_translator graphs `gold_study_to_database()`. Do we want to add it to the GOLD translation layer as well?

File: nmdc_runtime/site/ops.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pkalita-lbl) [reply nmdc_runtime/site/ops.py:755]: But that should be fine, right? If the structure of the input is the same, and the body of the method is the same I'm still not seeing why two methods are necessary.

File: nmdc_runtime/site/graphs.py (hunk 3 above)
Comment (pkalita-lbl) [reply nmdc_runtime/site/graphs.py]: Re: GOLD, yeah probably, but no real urgency to do that.

**Review by pkalita-lbl — APPROVED** [review]

File: nmdc_runtime/site/translation/neon_translator.py [hunk 6]
```diff
@@ -0,0 +1,819 @@
...
+
+        neon_mms_data_tables = (
+            "mms_metagenomeDnaExtraction",
+            "mms_metagenomeSequencing",
+        )
+
+        neon_sls_data_tables = (
```
Comment (mbthornton-lbl) [inline nmdc_runtime/site/translation/neon_translator.py:23]: These are getting defined at multiple places in the code - here and twice in repository.py. It would be better to define these kinds of things (keys, lists of tables to translate, etc.) once for a given data source

File: nmdc_runtime/site/translation/neon_translator.py [hunk 7]
```diff
@@ -0,0 +1,819 @@
...
+            omics_type=self._create_controlled_term_value(
+                omics_processing_row["investigation_type"].values[0]
+            ),
+            instrument_name=f"{self._get_value_or_none(omics_processing_row, 'sequencingMethod')} {self._get_value_or_none(omics_processing_row, 'instrument_model')}",
+            part_of="nmdc:sty-11-34xj1150",
+            name=f"Terrestrial soil microbial communities - {self._get_value_or_none(omics_processing_row, 'dnaSampleID')}",
+            type="nmdc:OmicsProcessing",
+        )
+
+    def get_database(self) -> nmdc.Database:
```
Comment (mbthornton-lbl) [inline nmdc_runtime/site/translation/neon_translator.py:372]: Consider a more descriptive function name to indicate that this returns the database with new data source translated and merged in.  Also consider docstrings for complex functions like this

File: nmdc_runtime/site/translation/neon_translator.py [hunk 8]
```diff
@@ -0,0 +1,819 @@
...
+                    mms_metagenomeDnaExtraction.processedDate,
+                    mms_metagenomeDnaExtraction.sampleMaterial,
+                    mms_metagenomeDnaExtraction.sampleMass,
+                    mms_metagenomeDnaExtraction.nucleicAcidConcentration,
+                    mms_metagenomeDnaExtraction.qaqcStatus
+                FROM mms_metagenomeSequencing
+                LEFT JOIN mms_metagenomeDnaExtraction ON mms_metagenomeDnaExtraction.dnaSampleID = mms_metagenomeSequencing.dnaSampleID
+            ) AS merged ON sls_metagenomicsPooling.genomicsSampleID = merged.genomicsSampleID
+        """
+        mms_sls_pooling_merged = pd.read_sql_query(query, self.conn)
```
Comment (mbthornton-lbl) [inline nmdc_runtime/site/translation/neon_translator.py:406]: It might improve readability to move these DB calls to sub-functions

File: tests/test_data/test_neon_data_translator.py [hunk 9]
```diff
@@ -0,0 +1,944 @@
...
+                "kclBlank3ID": "",
+                "kclBlank3Code": "",
+            },
+        ]
+    ),
+}
+
+
+class TestNeonDataTranslator:
+    @pytest.fixture
```
Comment (mbthornton-lbl) [inline tests/test_data/test_neon_data_translator.py:779]: Mock data above and test fixtures could be moved to conftest.py

File: nmdc_runtime/site/translation/neon_translator.py [hunk 10]
```diff
@@ -0,0 +1,819 @@
...
+
+        neon_raw_data_file_mappings_file = "https://raw.githubusercontent.com/microbiomedata/nmdc-schema/main/assets/misc/neon_raw_data_file_mappings.tsv"
+        self.neon_raw_data_file_mappings_df = pd.read_csv(
+            neon_raw_data_file_mappings_file, delimiter="\t"
+        )
+        self.neon_raw_data_file_mappings_df.to_sql(
+            "neonRawDataFile", self.conn, if_exists="replace", index=False
+        )
+
+    def _get_value_or_none(self, data, column_name):
```
Comment (mbthornton-lbl) [inline nmdc_runtime/site/translation/neon_translator.py:89]: Rename to reflect what this function does - normalizing NEON soil horizon terms

**Comment by mbthornton-lbl** [comment]
My overall comments:
 - consider a BaseTranslator class that implements common value(s) to nmdc.Object functions
 - long, complex functions can be more difficult to review and test. Consider breaking these up where possible
 - some classes / functions have doctrings, some do not. I would be good to come to some consensus as to what would be best here

File: nmdc_runtime/site/translation/neon_translator.py [hunk 11]
```diff
@@ -0,0 +1,819 @@
...
+            return None
+
+        return nmdc.GeolocationValue(
+            latitude=nmdc.DecimalDegree(latitude),
+            longitude=nmdc.DecimalDegree(longitude),
+        )
+
+    def _translate_biosample(self, neon_id, nmdc_id, biosample_row):
+        """
+        Translate a biosample_row DataFrame into a Biosample object.
```
[the remark this answers is not in the document]
Comment (turbomam) [reply nmdc_runtime/site/translation/neon_translator.py:178]: @cmungall always wants us to include RST docstrings but some of us are inconsistent.

[the remark this answers is not in the document]
Comment (turbomam) [reply nmdc_runtime/site/translation/neon_translator.py:178]: @sujaypatil96's more consistent than I am!

**Review by mbthornton-lbl — DISMISSED** [review]
Some of the refactoring I suggest here - e.g. moving common functions to a BaseTranslator class could be added to the work to build the related aquatic data translation / ingestion pipeline.

**Comment by sujaypatil96** [comment]
Apologies, I have no idea what "dismissed stale reviews" means, I've been playing around with various tool options in PyCharm. I thoroughly appreciate the code reviews from @pkalita-lbl and @mbthornton-lbl! ❤️

**Review by mbthornton-lbl — DISMISSED** [review]

File: .env.example [hunk 12]
```diff
@@ -31,4 +31,5 @@ GOLD_API_BASE_URL=https://gold.jgi.doe.gov/rest/nmdc
 GOLD_API_USERNAME=x
 GOLD_API_PASSWORD=x
 
-NMDC_PORTAL_API_BASE_URL=https://data-dev.microbiomedata.org/
+NEON_API_TOKEN=y
+NEON_API_BASE_URL=https://data.neonscience.org/api/v0
```
Comment (pkalita-lbl) [inline .env.example:35]: `NMDC_PORTAL_API_BASE_URL` intentionally removed?

**Review by pkalita-lbl — APPROVED** [review]

**Review by mbthornton-lbl — DISMISSED** [review]

File: .env.example (hunk 12 above)
Comment (sujaypatil96) [reply .env.example:37]: Oh wow, great catch! No, I removed it by mistake, apologies 😅

**Review by pkalita-lbl — APPROVED** [review]

**Comment by sujaypatil96** [comment]
@pkalita-lbl @mbthornton-lbl sorry 🥲 it looks like the way the repo is set up, if we add commits to the PR after review/approval it _dismisses_ the approval. Would you be able to help me with an _approval_ again? sorry.
